### ISSUE TITLE
Simplify the deserialization from VPack to DBEntity objects

### ISSUE THREAD

**Issue by christian-lechner** [arangodb/spring-data#34]
Hi!

Currently the `DefaultArangoConverter` must deal with non `DBEntity` objects because the `DBEntityDeserializer` only handles the root object. Nested objects are still "normal" hash maps and lists instead of `DBDocumentEntity` and `DBCollectionEntity`.

We could implement custom entity instantiators to produce `DBEntity` objects right away. As `DBDocumentEntity` and `DBCollectionEntity` are themselves a hash map and a collection we could do the following:
```java
public class DBEntityModule implements VPackModule {
	@Override
	public <C extends VPackSetupContext<C>> void setup(C context) {
		context.registerInstanceCreator(Map.class, new DBDocumentEntityInstantiator())
				.registerInstanceCreator(Collection.class, new DBCollectionEntityInstantiator());
	}
	public static class DBDocumentEntityInstantiator implements VPackInstanceCreator<Map<?, ?>> {
		@Override
		public Map<?, ?> createInstance() {
			return new DBDocumentEntity();
		}
	}
	public static class DBCollectionEntityInstantiator implements VPackInstanceCreator<Collection<?>> {
		@Override
		public Collection<?> createInstance() {
			return new DBCollectionEntity();
		}
	}
}
```

This would greatly simplify the `readMap()` and `readCollection()` methods of the converter.
E. g. new `readMap()` would look like the following:

```java
private Object readMap(final TypeInformation<?> type, final DBDocumentEntity source) {
	// ...
	for (final Map.Entry<String, Object> entry : source.entrySet()) {
		// ...
		if (value instanceof DBEntity) {
			map.put(key, read(valueType, (DBEntity) value));
		}
		// remove:
		// else if (value instanceof Map) {
		// map.put(key, read(valueType, new DBDocumentEntity((Map<? extends String, ? extends Object>) value)));
		// } else if (value instanceof Collection) {
		// map.put(key, read(valueType, new DBCollectionEntity((Collection<? extends Object>) value)));
		// } else if (isSimpleType(valueType.getType())) {
		// final Optional<Class<?>> customWriteTarget = conversions.getCustomWriteTarget(valueType.getType());
		// final Class<?> targetType = customWriteTarget.orElseGet(() -> valueType.getType());
		// map.put(key, conversionService.convert(value, targetType));
		// }
		else {
			/// TODO: potentially convert value
			map.put(key, value);
		}
	}
	return map;
}
```

Can you tell me why you ever needed this part in a read method:
```java
else if (isSimpleType(valueType.getType())) {
    final Optional<Class<?>> customWriteTarget = conversions.getCustomWriteTarget(valueType.getType());
    final Class<?> targetType = customWriteTarget.orElseGet(() -> valueType.getType());
    map.put(key, conversionService.convert(value, targetType));
}
```

**Comment by mpv1989** [comment]
Hi @christian-lechner,

good idea. We should change this like you suggested.

I don't remember why `else if (isSimpleType(valueType.getType())) {` is present. The tests run successful without it. So either we have not enough tests or this is not necessary.

best
Mark

**Comment by christian-lechner** [comment]
I can't think of a situation where this is necessary. Even if there is a conversion, it would be overwritten by the deserializing converters.

Then I will add this in my PR #33.

**Comment by christian-lechner** [comment]
Fixed by #33.
